### PR TITLE
feat: initial chainswaps

### PR DESCRIPTION
**Pull request by jackstar12**

### CODE REVIEW

File: boltz/api.go [hunk 1]
```diff
@@ -451,12 +552,65 @@ func (boltz *Boltz) CreateReverseSwap(request CreateReverseSwapRequest) (*Create
...
+	err := boltz.sendGetRequest(fmt.Sprintf("/v2/swap/chain/%s/claim", swapId), &response)
+
+	if response.Error != "" {
+		return nil, Error(errors.New(response.Error))
+	}
+
+	return &response, err
+}
+
+func (boltz *Boltz) SendChainSwapClaimSignature(swapId string, request *ChainSwapSigningRequest) (*PartialSignature, error) {
```
Comment (michael1011) [inline boltz/api.go:594]: Not a fan of the naming. You don't just send it but also get one in response. `ExchangeChainSwapClaimSignature`?

File: boltz/swaptree.go [hunk 2]
```diff
@@ -233,6 +234,10 @@ func checkScript(actual []byte, expected *txscript.ScriptBuilder) error {
 	}
 
 	if !bytes.Equal(actual, expectedScript) {
+		fmt.Println("expected")
+		fmt.Println(txscript.DisasmString(expectedScript))
+		fmt.Println("actual")
+		fmt.Println(txscript.DisasmString(actual))
```
Comment (michael1011) [inline boltz/swaptree.go:240]: rm

File: database/chain.go [hunk 3]
```diff
@@ -0,0 +1,429 @@
...
+	}
+
+	return data, nil
+}
+
+const refundableSwapsQuery = `
+SELECT swaps.*
+FROM chainSwaps swaps
+         JOIN chainSwapsData data ON swaps.id = data.id AND data.currency = swaps.fromCurrency
+WHERE (state = ? OR state = ? OR state = ?)
```
Comment (michael1011) [inline database/chain.go:280]:
```suggestion
WHERE (state = ? OR state = ? OR state = ?)
```
`IN` would be more elegant https://www.sqlitetutorial.net/sqlite-in/

File: database/chain.go [hunk 4]
```diff
@@ -0,0 +1,429 @@
...
+	return tx.Commit()
+}
+
+const insertChainSwapData = `
+		INSERT INTO chainSwapsData
+		(id, currency, privateKey, theirPublicKey, blindingKey, tree, amount, timeoutBlockheight, lockupTransactionId, transactionId, walletId, address, lockupAddress)
+		VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
+`
+
+func (database *Database) CreateChainSwapData(swapData ChainSwapData) error {
```
Comment (michael1011) [inline database/chain.go:168]: Why expose this?

File: database/chain.go [hunk 5]
```diff
@@ -0,0 +1,429 @@
...
+			return swap, err
+		}
+	} else {
+		return swap, errors.New("could not find Swap " + id)
+	}
+
+	return swap, err
+}
+
+func (database *Database) QueryChainSwapData(id string, currency boltz.Currency, isClaim bool) (data *ChainSwapData, err error) {
```
Comment (michael1011) [inline database/chain.go:252]: Same here: why expose this?

File: database/chain.go [hunk 6]
```diff
@@ -0,0 +1,429 @@
...
+	chainSwap.Error = error
+
+	_, err := database.Exec("UPDATE chainSwaps SET state = ?, error = ? WHERE id = ?", state, error, chainSwap.Id)
+	return err
+}
+
+func (database *Database) QueryChainSwap(id string) (swap *ChainSwap, err error) {
+	database.lock.RLock()
+	defer database.lock.RUnlock()
+	rows, err := database.Query("SELECT * FROM chainSwaps WHERE id = ?", id)
```
Comment (michael1011) [inline database/chain.go:232]: Should only return a single row `QueryRow`

File: boltzrpc/boltzrpc.proto [hunk 7]
```diff
@@ -28,6 +28,11 @@ service Boltz {
     */
     rpc GetReversePair (Pair) returns (ReversePair);
 
+    /*
+    Fetches information about a specific pair for a reverse swap.
```
Comment (michael1011) [inline boltzrpc/boltzrpc.proto:32]:
```suggestion
    Fetches information about a specific pair for a chain swap.
```

File: nursery/chain.go [hunk 8]
```diff
@@ -0,0 +1,231 @@
...
+}
+
+func (nursery *Nursery) setChainSwapLockupTransaction(swap *database.ChainSwap, data *database.ChainSwapData, transactionId string) error {
+	data.LockupTransactionId = transactionId
+	_, _, value, err := nursery.findLockupVout(data)
+	if err != nil {
+		return fmt.Errorf("could not find lockup vout: %s", err)
+	}
+
+	logger.Info("Got lockup transaction of Swap " + data.Id + " from Boltz: " + transactionId)
```
Comment (michael1011) [inline nursery/chain.go:39]: One thing I am wondering about when looking at these logs: should we prefix those `Swap` with the type? Like `Chain Swap`? That's how I do it in the backend and that makes it quite a bit easier to follow along when multiple swaps are going on

File: rpcserver/router.go [hunk 9]
```diff
@@ -687,6 +689,200 @@ func (server *routedBoltzServer) CreateReverseSwap(ctx context.Context, request
...
+			Address:            response.Address,
+			ExpectedAmount:     int64(response.ExpectedAmount),
+			Bip21:              response.Bip21,
+			TimeoutBlockHeight: response.TimeoutBlockHeight,
+			TimeoutHours:       float32(timeoutHours),
+		}
+	*/
+
+	if !request.GetExternalPay() {
+		// TODO: custom block target?
```
Comment (michael1011) [inline rpcserver/router.go:858]: Maybe also custom fee? Or do we leave all those custom options to the user via external pay?

File: nursery/refund.go [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+		currency = chainSwap.Pair.From
+		refundOutput, err := nursery.getChainSwapRefundOutput(&chainSwap)
+		if err != nil {
+			return fmt.Errorf("could not get refund output for chain swap %s: %v", chainSwap.Id, err)
+		}
+
+		refundOutput.Cooperative = cooperative
+		refundedChainSwaps = append(refundedChainSwaps, chainSwap)
+		outputs = append(outputs, refundOutput)
+	}
```
Comment (michael1011) [inline nursery/refund.go:70]: Any chance we could make them adhere to a common interface to avoid the code duplication?

File: boltz/api.go [hunk 11]
```diff
@@ -379,12 +480,12 @@ func (boltz *Boltz) CreateSwap(request CreateSwapRequest) (*CreateSwapResponse,
 	return &response, err
 }
 
-func (boltz *Boltz) RefundSwap(request RefundSwapRequest) (*PartialSignature, error) {
+func (boltz *Boltz) RefundSwap(id string, request *RefundRequest) (*PartialSignature, error) {
```
Comment (michael1011) [inline boltz/api.go:483]:
```suggestion
func (boltz *Boltz) RefundSwap(swapId string, request *RefundRequest) (*PartialSignature, error) {
```

To be consistent

File: boltz/api.go [hunk 12]
```diff
@@ -451,12 +552,65 @@ func (boltz *Boltz) CreateReverseSwap(request CreateReverseSwapRequest) (*Create
 	return &response, err
 }
 
-func (boltz *Boltz) ClaimReverseSwap(request ClaimReverseSwapRequest) (*PartialSignature, error) {
+func (boltz *Boltz) ClaimReverseSwap(id string, request *ClaimRequest) (*PartialSignature, error) {
```
Comment (michael1011) [inline boltz/api.go:555]:
```suggestion
func (boltz *Boltz) ClaimReverseSwap(swapId string, request *ClaimRequest) (*PartialSignature, error) {
```

File: boltz/swaptree.go [hunk 13]
```diff
@@ -233,6 +242,10 @@ func checkScript(actual []byte, expected *txscript.ScriptBuilder) error {
...
```
Comment (michael1011) [inline boltz/swaptree.go:248]: rm

File: boltzrpc/boltzrpc.proto [hunk 14]
```diff
@@ -497,6 +530,58 @@ message CreateReverseSwapResponse {
     optional string invoice = 5;
 }
 
+message CreateChainSwapRequest {
+    // amount of satoshis to swap
+    uint64 amount = 1;
```
Comment (michael1011) [inline boltzrpc/boltzrpc.proto:535]: Is this the equivalent to `userLockAmount`? Not clear from the comment

File: boltzrpc/boltzrpc.proto [hunk 15]
```diff
@@ -497,6 +530,58 @@ message CreateReverseSwapResponse {
...
+    string id = 1;
+    Currency currency = 2;
+    string private_key = 3;
+    string their_public_key = 4;
+    int64 amount = 6;
+    uint32 timeout_block_height = 7;
+    string lockup_transaction_id = 8;
+    string transaction_id = 9;
+    string wallet = 11;
+    string address = 12;
```
Comment (michael1011) [inline boltzrpc/boltzrpc.proto:580]: Is it possible that there is no address yet when a wallet and no address is set in the request to create?

File: nursery/refund.go [hunk 16]
```diff
@@ -0,0 +1,70 @@
...
+			}
+
+			chainSwaps, err := nursery.database.QueryRefundableChainSwaps(currency)
+			if err != nil {
+				logger.Error("Could not query refundable Swaps: " + err.Error())
+				continue
+			}
+
+			if len(swaps) > 0 || len(chainSwaps) > 0 {
+				logger.Infof("Found %d Swaps to refund at height %d", len(swaps), newBlock.Height)
```
Comment (michael1011) [inline nursery/refund.go:31]:
```suggestion
				logger.Infof("Found %d Swaps to refund at height %d", len(swaps) + len(chainSwaps), newBlock.Height)
```

File: rpcserver/router.go [hunk 17]
```diff
@@ -687,6 +729,209 @@ func (server *routedBoltzServer) CreateReverseSwap(ctx context.Context, request
...
+		EntityId: macaroons.EntityFromContext(ctx),
+	})
+	externalPay := request.GetExternalPay()
+	if err != nil && !externalPay {
+		// the specified name was wrong
+		if fromName != "" {
+			return nil, handleError(err)
+		}
+		// no name was specified, which means there is no wallet available
+		externalPay = true
```
Comment (michael1011) [inline rpcserver/router.go:793]: Not sure if we want to enable this implicitly

File: rpcserver/router.go [hunk 18]
```diff
@@ -687,6 +729,209 @@ func (server *routedBoltzServer) CreateReverseSwap(ctx context.Context, request
...
+
+	if err := server.nursery.RegisterChainSwap(chainSwap); err != nil {
+		return nil, handleError(err)
+	}
+
+	/*
+		if err := server.nursery.RegisterSwap(chainSwap); err != nil {
+			return nil, handleError(err)
+		}
+	*/
```
Comment (michael1011) [inline rpcserver/router.go:930]: rm

File: rpcserver/router.go [hunk 19]
```diff
@@ -687,6 +729,209 @@ func (server *routedBoltzServer) CreateReverseSwap(ctx context.Context, request
...
+		timeoutHours := boltz.BlocksToHours(response.TimeoutBlockHeight-blockHeight, pair.From)
+		ChainSwapResponse := &boltzrpc.CreateChainSwapResponse{
+			Id:                 chainSwap.Id,
+			Address:            response.Address,
+			ExpectedAmount:     int64(response.ExpectedAmount),
+			Bip21:              response.Bip21,
+			TimeoutBlockHeight: response.TimeoutBlockHeight,
+			TimeoutHours:       float32(timeoutHours),
+		}
+	*/
```
Comment (michael1011) [inline rpcserver/router.go:904]: rm

File: rpcserver/router.go [hunk 20]
```diff
@@ -687,6 +729,209 @@ func (server *routedBoltzServer) CreateReverseSwap(ctx context.Context, request
...
+		return swapData, nil
+	}
+
+	chainSwap.ToData, err = parseDetails(response.ClaimDetails, pair.To)
+	if err != nil {
+		return nil, handleError(err)
+	}
+	if toWallet != nil {
+		id := toWallet.GetWalletInfo().Id
+		chainSwap.ToData.WalletId = &id
```
Comment (michael1011) [inline rpcserver/router.go:875]: Maybe I missed it but where do we set the claim address if one was specified in the request?

File: boltzrpc/boltzrpc.proto (hunk 15 above)
Comment (jackstar12) [reply boltzrpc/boltzrpc.proto:580]: Yes. Should we make those optional?

File: boltzrpc/boltzrpc.proto (hunk 14 above)
Comment (jackstar12) [reply boltzrpc/boltzrpc.proto:535]: Yes. Should we expose both user and server lock amount in the api?

File: rpcserver/router.go (hunk 20 above)
Comment (jackstar12) [reply rpcserver/router.go:875]: Here: https://github.com/BoltzExchange/boltz-client/pull/133/files#diff-3c0751abe8a29668b3f795bfe68a1a0dd18a58e1b76816afb7e4ce0d79cc28bcR836

File: boltzrpc/boltzrpc.proto (hunk 15 above)
Comment (michael1011) [reply boltzrpc/boltzrpc.proto:580]: Yes

File: boltzrpc/boltzrpc.proto (hunk 14 above)
Comment (michael1011) [reply boltzrpc/boltzrpc.proto:535]: Our just write in the comment that it is the amount you are expected to send

**Review by kilrau — CHANGES_REQUESTED** [review]
Ignore

**Review by kilrau — CHANGES_REQUESTED** [review]
As just discussed, db migration is still missing

**Review by kilrau — APPROVED** [review]
db migration works now (implemented in https://github.com/BoltzExchange/boltz-client/pull/136/commits/75dfa6fc22e30b6467d124a06aaead6e05cd3067 in a different branch). Let's merge this one if good with you @michael1011

File: boltzrpc/boltzrpc.proto [hunk 21]
```diff
@@ -28,19 +28,24 @@ service Boltz {
...
+    rpc GetChainPair (Pair) returns (ChainPair);
+
     /*
     Fetches all available pairs for submarine and reverse swaps.
     */
     rpc GetPairs (google.protobuf.Empty) returns (GetPairsResponse);
 
     /*
-    Returns a list of all swaps, reverse swaps and channel creations in the database.
+    Returns a list of all swaps, reverse swaps, and chain swaps n the database.
```
Comment (michael1011) [inline boltzrpc/boltzrpc.proto:42]:
```suggestion
    Returns a list of all swaps, reverse swaps, and chain swaps in the database.
```

File: boltzrpc/boltzrpc.proto [hunk 22]
```diff
@@ -28,19 +28,24 @@ service Boltz {
...
     */
     rpc ListSwaps (ListSwapsRequest) returns (ListSwapsResponse);
 
     /*
-    Refund a failed swap manually. 
-    This is only required when no refund address has been set or the daemon has no wallet for the currency.
+    Refund a failed swap manually.
+    This is only required when no refund address has been set or the the swap does not have an associated wallet.
```
Comment (michael1011) [inline boltzrpc/boltzrpc.proto:48]:
```suggestion
    This is only required when no refund address has been set or the swap does not have an associated wallet.
```

File: boltzrpc/boltzrpc.proto [hunk 23]
```diff
@@ -368,9 +379,30 @@ message ReversePair {
...
+        }
+        message Miner {
+            uint64 server = 1;
+            User user = 2;
+        }
+        float percentage = 1;
+        Miner miner_fees = 2;
+    }
+    Fees fees = 5;
+}
```
Comment (michael1011) [inline boltzrpc/boltzrpc.proto:400]:
```suggestion
message ChainPair {
    message Fees {
        message User {
            uint64 lockup = 1;
            uint64 claim = 2;
        }
        message Miner {
            uint64 server = 1;
            User user = 2;
        }
        float percentage = 1;
        Miner miner_fees = 2;
    }

    Pair pair = 1;
    string hash = 2;
    float rate = 3;
    Limits limits = 4;
    Fees fees = 5;
}
```

Looks cleaner imho

File: boltzrpc/boltzrpc.proto [hunk 24]
```diff
@@ -497,6 +530,58 @@ message CreateReverseSwapResponse {
     optional string invoice = 5;
 }
 
+message CreateChainSwapRequest {
+    // amount of satoshis to swap. it will be the amount expected to be sent to the lockup address.
```
Comment (michael1011) [inline boltzrpc/boltzrpc.proto:534]:
```suggestion
    // Amount of satoshis to swap. It is the amount expected to be sent to the lockup address.
```

File: boltzrpc/boltzrpc.proto [hunk 25]
```diff
@@ -497,6 +530,58 @@ message CreateReverseSwapResponse {
...
+    uint64 amount = 1;
+    Pair pair = 2;
+    // address where funds will go if the swap succeeds
```
Comment (michael1011) [inline boltzrpc/boltzrpc.proto:537]:
```suggestion
    // Address where funds will be swept to if the swap succeeds
```

File: boltzrpc/boltzrpc.proto [hunk 26]
```diff
@@ -497,6 +530,58 @@ message CreateReverseSwapResponse {
...
+    optional string to_address = 3;
+    // address where the coins should go if the swap fails.
+    optional string refund_address = 4;
+    // Wallet from which the swap should be paid from. Ignored if `external_pay` is set to true.
+    // If the swap fails, funds will be refunded to this wallet aswell.
+    optional int64 from_wallet_id = 5;
+    // wallet where the the funds will go if the swap succeeds.
```
Comment (michael1011) [inline boltzrpc/boltzrpc.proto:544]:
```suggestion
    // Wallet where the the funds will go if the swap succeeds.
```

File: boltzrpc/boltzrpc.proto [hunk 27]
```diff
@@ -497,6 +530,58 @@ message CreateReverseSwapResponse {
...
+    optional int64 to_wallet_id = 6;
+    // Whether the daemon should broadcast the claim transaction immediately after the lockup transaction is in the mempool.
+    // Should only be used for smaller amounts as it involves trust in boltz.
```
Comment (michael1011) [inline boltzrpc/boltzrpc.proto:547]:
```suggestion
    // Should only be used for smaller amounts as it involves trust in Boltz.
```

File: rpcserver/router.go [hunk 28]
```diff
@@ -63,6 +63,40 @@ func handleError(err error) error {
...
+	heights *boltzrpc.BlockHeights, swaps []database.Swap, chainSwaps []database.ChainSwap, err error,
+) {
+	heights = &boltzrpc.BlockHeights{}
+	heights.Btc, err = server.onchain.GetBlockHeight(boltz.CurrencyBtc)
+	if err != nil {
+		return
+	}
+	liquidHeight, err := server.onchain.GetBlockHeight(boltz.CurrencyLiquid)
+	if err != nil {
+		logger.Infof("Failed to get block height for liquid: %v", err)
```
Comment (michael1011) [inline rpcserver/router.go:76]: Should probably be a warning

File: rpcserver/router.go [hunk 29]
```diff
@@ -89,14 +123,9 @@ func (server *routedBoltzServer) GetInfo(ctx context.Context, _ *boltzrpc.GetInf
...
-	refundableSwaps, err := server.database.QueryRefundableSwaps("")
+	blockHeights, refundableSwaps, refundableChainSwaps, err := server.queryRefundableSwaps()
 	if err != nil {
-		return nil, handleError(err)
-	}
-
-	refundableChainSwaps, err := server.database.QueryRefundableChainSwaps("")
-	if err != nil {
-		return nil, handleError(err)
+		return nil, handleError(fmt.Errorf("Failed to get block height for btc: %v", err))
```
Comment (michael1011) [inline rpcserver/router.go:128]: "Failed to fetch refundable swaps: %v", err
?

**Review by michael1011 — APPROVED** [review]
Minor wording things. Can be merged afterwards

**Pull request merged**